strawberryfields/ops.py

```diff
@@ -1808,6 +1808,98 @@ def _decompose(self, reg, **kwargs):
         return cmds
 
 
+class BipartiteGraphEmbed(Decomposition):
+    r"""Embed a bipartite graph into an interferometer setup.
+
+    A bipartite graph is a graph that consists of two vertex sets :math:`U` and :math:`V`,
+    such that every edge in the graph connects a vertex between :math:`U` and :math:`V`.
+    That is, there are no edges between vertices in the same vertex set.
+
+    The adjacency matrix of an :math:`N` vertex undirected bipartite graph
+    is a :math:`N\times N` symmetric matrix of the form
+
+    .. math:: A = \begin{bmatrix}0 & B \\ B^T & 0\end{bmatrix}
+
+    where :math:`B` is a :math:N/2\times N/2` matrix representing the (weighted)
+    edges between the vertex set.
+
+    This operation uses the Takagi decomposition to decompose
+    an adjacency matrix into a sequence of two mode squeezers, beamsplitters, and
+    rotation gates.
+
+    Args:
+        A (array): Either an :math:`N\times N` complex or real symmetric adjacency matrix
+            :math:`A`, or an :math:`N/2\times N/2` complex or real matrix :math:`B`
+            representing the edges between the vertex sets if ``edges=True``.
+        mean_photon (float): guarantees that the mean photon number in the pure Gaussian state
+            representing the graph satisfies  :math:`\frac{1}{N}\sum_{i=1}^N sinh(r_{i})^2 ==` :code:``mean_photon``
+        edges (bool): set to ``True`` if argument ``A`` represents the edges :math:`B`
+            between the vertex sets rather than the full adjacency matrix
+        drop_identity (bool): If ``True``, decomposed gates with trivial parameters,
+            such that they correspond to an identity operation, are removed.
+        tol (float): the tolerance used when checking if the input matrix is symmetric:
+            :math:`|A-A^T| <` tol
+    """
+
+    def __init__(self, A, mean_photon=1.0, edges=False, drop_identity=True, tol=1e-6):
+        self.mean_photon = mean_photon
+        self.tol = tol
+        self.identity = np.all(np.abs(A - np.identity(len(A))) < _decomposition_merge_tol)
+        self.drop_identity = drop_identity
+
+        if edges:
+            self.ns = 2*A.shape[0]
+            B = A
+        else:
+            self.ns = A.shape[0]
+
+            # check if A is a bipartite graph
+            N = A.shape[0]//2
+            A00 = A[:N, :N]
+            A11 = A[N:, N:]
+
+            diag_zeros = np.allclose(A00, np.zeros_like(A00), atol=tol, rtol=0)
+            diag_zeros = diag_zeros and np.allclose(A11, np.zeros_like(A11), atol=tol, rtol=0)
+
+            if (not diag_zeros) or (not np.allclose(A, A.T, atol=tol, rtol=0)):
+                raise ValueError("Adjacency matrix {} does not represent a bipartite graph".format(A))
+
+            B = A[:N, N:]
+
+        super().__init__([B])
+
+    def _decompose(self, reg, **kwargs):
```

Comment:
Can a docstring be added here?

Reply:
Since this is an inherited method of an abstract method, my preference is not to, as it will automatically inherit the docstring of the parent method. This avoids duplication, and the issue of updating one `_decompose()` method, forgetting others, and having them go out-of-sync.

We used to have docstrings for each of these methods, but there was a PR to remove them as we kept running into the sync issue. (I think this was when a particular python/sphinx version was released that supported inheriting docstrings?)

Reply:
Ah makes sense, thanks!